Refactor get or load program

janezpodhostnik opened the pull request:
Refactor GetAndSet (a.k.a. [GetOrLoad](https://github.com/onflow/cadence/pull/2362)) to make it easier to understand what is going on.

The next step is to remove the individual get and set methods, as they shouldn't be used anymore.

Work towards: https://github.com/onflow/cadence/issues/1684

pattyshack reviewed (changes requested):
could you rename the variables?  single letter variables aren't very readable

janezpodhostnik commented:
> could you rename the variables? single letter variables aren't very readable

Renamed, and fixed test. I hat to move the freeze check before the `get` part. 

The reason is that the freeze check touches registers, so it has to be done either the program is in cache or not.
The other solution to this is to put the freeze check into the nested state transaction, but I'll do that in the next PR.

pattyshack reviewed (approved):
thx for renaming

zhangchiqing commented on fvm/environment/programs.go:
```diff
@@ -190,31 +190,166 @@ func (programs *Programs) GetAndSetProgram(
+	if programState.BytesWritten() > 0 {
+		// This should never happen. Loading a program should not write to the state.
+		// If this happens, it indicates an implementation error.
+		return nil, nil, fmt.Errorf("cannot set program. State was written to during program parsing")
```
```suggestion
		return nil, nil, fmt.Errorf("cannot set program to address %v. State was written to during program parsing", address)
```

janezpodhostnik commented on fvm/environment/programs.go:
```diff
@@ -190,31 +190,166 @@ func (programs *Programs) GetAndSetProgram(
 ) (*interpreter.Program, error) {
+	// ignore empty locations
+	if location == nil {
+		return nil, nil
```
I'm not actually sure why this is here. I kept it during the refactor as I did not want to change any logic to much. 

I will investigate why this is here and if we can remove it in a separate PR.

zhangchiqing replied:
OK, maybe adding a TODO to remove it.

janezpodhostnik commented on fvm/environment/programs.go:
```diff
@@ -190,31 +190,166 @@ func (programs *Programs) GetAndSetProgram(
 	location common.Location,
 	load func() (*interpreter.Program, error),
 ) (*interpreter.Program, error) {
+	// ignore empty locations
+	if location == nil {
+		return nil, nil
+	}
```
see [above](https://github.com/onflow/flow-go/pull/3976/files#r1124297441)

janezpodhostnik commented:
bors merge

janezpodhostnik commented:
bors merge

janezpodhostnik commented:
bors merge
